File: scripts/generate-market-data.py
#!/usr/bin/env python3
"""
generate_market_data.py

Produces a deterministic, reproducible synthetic order-flow dataset for
replay against OrderBook in simulation.cpp. This is a one-off tool (run
once, produce a CSV, done) so it's written in Python for readability
rather than C++ -- the generator's implementation language is invisible
to everything downstream, since simulation.cpp only ever reads the CSV.

Usage:
    python3 generate_market_data.py <num_ticks> <output.csv> [seed]

Output format (CSV, one line per operation):
    op,id,side,type,price,qty,tick,sequence
        op        NEW | CANCEL
        id        order id
        side      BUY | SELL          (empty for CANCEL)
        type      LIMIT | MARKET      (empty for CANCEL)
        price     integer price       (empty for CANCEL, empty for MARKET NEW)
        qty       integer quantity    (empty for CANCEL)
        tick      the simulation tick this operation was generated in.
                  Multiple rows can share a tick (a burst tick emits many
                  orders at once); ticks can also be skipped entirely (a
                  quiet tick can emit zero orders). This is what carries
                  the regime/clustering signal.
        sequence  monotonically increasing row counter, unique per row.
                  Used only for tie-breaking; replay order is already
                  determined by file order, not by this column.

Design notes:
    - Price offsets drawn from a Laplace distribution around a slowly
      random-walking mid price -> realistic clustering near the touch
      with fat tails, rather than uniform price selection.
    - Order sizes drawn from a log-normal distribution -> many small
      orders, occasional large ones, matching real order-size skew.
    - A two-state Markov chain (QUIET / BURST) drives arrival intensity
      and side imbalance over time, producing bursty, directionally-
      biased flow instead of a flat, balanced Poisson process.
    - A fraction of resting limit orders are scheduled for cancellation
      some number of ticks after insertion, so CANCEL and the
      empty-price-level cleanup path both get realistic exercise
      downstream. Scheduled cancels are kept in a min-heap ordered by
      fire_at_tick (not a plain FIFO queue) since the delay is
      randomized -- a far-future cancel can be scheduled before a
      near-future one, so only a priority queue guarantees the
      soonest-due cancel always fires first.
"""

import csv
import heapq
import sys
from dataclasses import dataclass, field
from enum import Enum, auto


class Regime(Enum):
    QUIET = auto()
    BURST = auto()


# Regime transition probabilities (per tick).
QUIET_TO_BURST = 0.01  # ~1 in 100 ticks
BURST_TO_QUIET = 0.05  # bursts last ~20 ticks on average


@dataclass
class SimState:
    rng: "numpy.random.Generator"
    mid_price: float = 10000.0          # integer-price ticks, e.g. cents
    regime: Regime = Regime.QUIET
    side_bias: float = 0.0              # -1..+1, skews P(buy) during bursts
    next_id: int = 1
    sequence: int = 0                   # monotonic row counter, not a time value
    pending_cancels: list = field(default_factory=list)  # min-heap: (fire_at_tick, id)


def update_regime(state: SimState) -> None:
    """Maybe flip QUIET <-> BURST this tick; entering BURST picks a fresh
    momentum direction/magnitude for side_bias."""
    if state.regime is Regime.QUIET:
        if state.rng.random() < QUIET_TO_BURST:
            state.regime = Regime.BURST
            direction = -1.0 if state.rng.random() < 0.5 else 1.0
            state.side_bias = direction * (0.2 + 0.3 * state.rng.random())
    else:
        if state.rng.random() < BURST_TO_QUIET:
            state.regime = Regime.QUIET
            state.side_bias = 0.0


def update_mid_price(state: SimState) -> None:
    """One random-walk step, nudged by side_bias during a directional burst.
    Must run after update_regime (reads the fresh side_bias) and before
    generate_arrivals (which prices orders off mid_price)."""
    step = state.rng.normal(0.0, 1.5)
    state.mid_price += step + state.side_bias * 0.8
    state.mid_price = max(state.mid_price, 100.0)  # keep prices positive


def fire_due_cancels(state: SimState, writer, tick: int) -> None:
    """Drain every cancel scheduled for exactly this tick."""
    while state.pending_cancels and state.pending_cancels[0][0] == tick:
        _, order_id = heapq.heappop(state.pending_cancels)
        writer.writerow(["CANCEL", order_id, "", "", "", "", tick, state.sequence])
        state.sequence += 1


def generate_arrivals(state: SimState, writer, tick: int) -> None:
    """Generate this tick's new order arrivals: how many, then for each
    one, side/type/size/price, emit it, and maybe schedule a future
    cancel."""
    mean_arrivals = 2 if state.regime is Regime.QUIET else 8
    n_arrivals = state.rng.poisson(mean_arrivals)

    p_buy = min(max(0.5 + state.side_bias * 0.5, 0.05), 0.95)

    for _ in range(n_arrivals):
        order_id = state.next_id
        state.next_id += 1

        buy = state.rng.random() < p_buy
        market = state.rng.random() < 0.08  # ~8% of flow is market orders

        qty = max(1, round(state.rng.lognormal(3.0, 0.9)))

        if market:
            side = "BUY" if buy else "SELL"
            writer.writerow(["NEW", order_id, side, "MARKET", "", qty, tick, state.sequence])
            state.sequence += 1
            continue

        offset = state.rng.laplace(0.0, 4.0)
        # Bias resting limit orders to sit on their own side of the mid
        # (buys below mid, sells above mid) so the book doesn't
        # self-cross on arrival more than realistically expected.
        signed_offset = -abs(offset) if buy else abs(offset)
        price = max(1, round(state.mid_price) + round(signed_offset))

        side = "BUY" if buy else "SELL"
        writer.writerow(["NEW", order_id, side, "LIMIT", price, qty, tick, state.sequence])
        state.sequence += 1

        if state.rng.random() < 0.35:  # ~35% of limit orders get a scheduled cancel
            delay = state.rng.integers(5, 201)  # [5, 200] ticks later
            heapq.heappush(state.pending_cancels, (tick + delay, order_id))


def flush_remaining_cancels(state: SimState, writer) -> None:
    """Flush any cancels still pending past the last tick, so scheduled
    cancels aren't silently dropped from the dataset. The heap pops them
    in fire_at_tick order."""
    while state.pending_cancels:
        fire_at_tick, order_id = heapq.heappop(state.pending_cancels)
        writer.writerow(["CANCEL", order_id, "", "", "", "", fire_at_tick, state.sequence])
        state.sequence += 1


def main() -> int:
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} <num_ticks> <output.csv> [seed]", file=sys.stderr)
        return 1

    num_ticks = int(sys.argv[1])
    output_path = sys.argv[2]
    seed = int(sys.argv[3]) if len(sys.argv) >= 4 else 42

    import numpy as np  # local import so --help-style usage errors don't require numpy

    state = SimState(rng=np.random.default_rng(seed))

    with open(output_path, "w", newline="") as f:
        writer = csv.writer(f, lineterminator="\n")
        writer.writerow(["op", "id", "side", "type", "price", "qty", "tick", "sequence"])

        for tick in range(num_ticks):
            update_regime(state)
            update_mid_price(state)
            fire_due_cancels(state, writer, tick)
            generate_arrivals(state, writer, tick)

        flush_remaining_cancels(state, writer)

    print(f"Generated {state.next_id - 1} orders across {num_ticks} ticks "
          f"-> {output_path} (seed={seed})", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: generate-market-data.cpp
// generate_market_data.cpp
//
// Produces a deterministic, reproducible synthetic order-flow dataset for
// replay against OrderBook in simulation.cpp. Designed to be single-sourced:
// the same dataset feeds both the invariant-checking loop and the
// performance-profiling loop, so results from the two are comparable and
// bugs found in one are reproducible in the other.
//
// Usage:
//   ./generate_market_data <num_ticks> <output.csv> [seed]
//
// Output format (CSV, one line per operation):
//   op,id,side,type,price,qty,timestamp
//     op        NEW | CANCEL
//     id        order id (uint64)
//     side      BUY | SELL          (empty for CANCEL)
//     type      LIMIT | MARKET      (empty for CANCEL)
//     price     integer price       (empty for CANCEL, empty for MARKET NEW)
//     qty       integer quantity    (empty for CANCEL)
//     timestamp monotonically increasing tick/sequence index
//
// Design notes:
//   - Price offsets drawn from a discretized Laplace distribution around a
//     slowly random-walking mid price -> realistic clustering near the touch
//     with fat tails, rather than uniform price selection.
//   - Order sizes drawn from a log-normal distribution -> many small orders,
//     occasional large ones, matching real order-size skew.
//   - A two-state Markov chain (QUIET / BURST) drives arrival intensity and
//     side imbalance over time, producing bursty, directionally-biased flow
//     instead of a flat, balanced Poisson process.
//   - A fraction of resting orders are scheduled for cancellation some
//     number of ticks after insertion, so CANCEL and the empty-level
//     cleanup path both get realistic exercise downstream. Scheduled
//     cancels are kept in a min-heap ordered by fire_at_tick (not a plain
//     FIFO queue) since cancel_delay is randomized -- a far-future cancel
//     can be scheduled before a near-future one, so only a priority queue
//     guarantees the soonest-due cancel always fires first.
//
// Structure: simulation state and distribution objects are grouped into
// SimState / Distributions so each tick-phase can be its own function with
// a self-documenting signature, rather than one long main() closing over a
// dozen loose locals via lambda captures.

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <queue>
#include <random>
#include <string>
#include <vector>

using OrderId = std::uint64_t;
using Price = std::int64_t;
using Qty = std::int64_t;

enum class Regime { Quiet, Burst };

struct PendingCancel {
  OrderId id;
  std::uint64_t fire_at_tick;
};

// Order by fire_at_tick so std::priority_queue (a max-heap by default)
// surfaces the SOONEST-due cancel at the top when used with this comparator
// (see the priority_queue member below, which this flips into a min-heap).
struct PendingCancelLater {
  bool operator()(const PendingCancel &a, const PendingCancel &b) const {
    return a.fire_at_tick > b.fire_at_tick;
  }
};

namespace {

// Draw from a discretized Laplace distribution centered at 0 with the given
// scale (b). Laplace has fatter tails than Gaussian, which better matches
// how order book depth actually falls off away from the touch. <random>
// has no built-in laplace_distribution, so this is the standard
// inverse-CDF construction on top of uniform_real_distribution.
Price sample_laplace_offset(std::mt19937_64 &rng, double scale) {
  std::uniform_real_distribution<double> u(-0.5, 0.5);
  double x = u(rng);
  double sign = (x < 0.0) ? -1.0 : 1.0;
  double magnitude = -scale * std::log(1.0 - 2.0 * std::abs(x));
  return static_cast<Price>(std::llround(sign * magnitude));
}

} // namespace

// ---------------------------------------------------------------------
// Simulation state: everything that evolves tick-by-tick.
// ---------------------------------------------------------------------
struct SimState {
  std::mt19937_64 rng;

  double mid_price = 10000.0; // integer-price ticks, e.g. cents
  Regime regime = Regime::Quiet;
  double side_bias = 0.0; // -1..+1, skews P(buy) away from 0.5 during bursts

  OrderId next_id = 1;
  std::uint64_t timestamp = 0;

  // Min-heap ordered by fire_at_tick -- see PendingCancelLater above.
  std::priority_queue<PendingCancel, std::vector<PendingCancel>,
                      PendingCancelLater>
      pending_cancels;

  explicit SimState(std::uint64_t seed) : rng(seed) {}
};

// ---------------------------------------------------------------------
// Distribution objects: fixed shape for the whole run, parameters chosen
// once up front. Grouped here so each tick-phase function can take a
// single `dist` argument instead of half a dozen individual ones.
// ---------------------------------------------------------------------
struct Distributions {
  // Mid-price random walk step.
  std::normal_distribution<double> mid_walk{0.0, 1.5};

  // Log-normal order size: mean/stddev chosen so typical sizes land in a
  // realistic small-to-moderate range with an occasional large outlier.
  std::lognormal_distribution<double> size_dist{3.0, 0.9};

  // General-purpose uniform draw, used for regime transition rolls and
  // burst-direction/magnitude selection.
  std::uniform_real_distribution<double> unit{0.0, 1.0};

  // Orders per tick, regime-dependent.
  std::poisson_distribution<int> quiet_arrivals{2};
  std::poisson_distribution<int> burst_arrivals{8};

  // Order type mix: most flow is limit orders.
  std::bernoulli_distribution is_market{0.08};

  // Cancellation: fraction of NEW limit orders get scheduled for a later
  // cancel, at a randomly chosen delay.
  std::bernoulli_distribution will_cancel{0.35};
  std::uniform_int_distribution<int> cancel_delay{5, 200};
};

// Regime transition probabilities (per tick).
constexpr double QUIET_TO_BURST = 0.01; // ~1 in 100 ticks
constexpr double BURST_TO_QUIET = 0.05; // bursts last ~20 ticks on avg

// ---------------------------------------------------------------------
// Output formatting
// ---------------------------------------------------------------------

void emit_new(std::ofstream &out, OrderId id, bool buy, bool market,
              Price price, Qty qty, std::uint64_t ts) {
  out << "NEW," << id << ',' << (buy ? "BUY" : "SELL") << ','
      << (market ? "MARKET" : "LIMIT") << ',';
  if (!market) {
    out << price;
  }
  out << ',' << qty << ',' << ts << '\n';
}

void emit_cancel(std::ofstream &out, OrderId id, std::uint64_t ts) {
  out << "CANCEL," << id << ",,,," << ',' << ts << '\n';
}

// ---------------------------------------------------------------------
// Per-tick phases
// ---------------------------------------------------------------------

// Maybe flip Quiet <-> Burst this tick, and if entering Burst, pick a
// fresh momentum direction/magnitude for side_bias.
void update_regime(SimState &state, Distributions &dist) {
  if (state.regime == Regime::Quiet) {
    if (dist.unit(state.rng) < QUIET_TO_BURST) {
      state.regime = Regime::Burst;
      state.side_bias = (dist.unit(state.rng) < 0.5 ? -1.0 : 1.0) *
                        (0.2 + 0.3 * dist.unit(state.rng));
    }
  } else {
    if (dist.unit(state.rng) < BURST_TO_QUIET) {
      state.regime = Regime::Quiet;
      state.side_bias = 0.0;
    }
  }
}

// One random-walk step for mid_price, nudged by side_bias if we're in a
// directional burst. Must run after update_regime (reads the fresh
// side_bias) and before generate_arrivals (which prices off mid_price).
void update_mid_price(SimState &state, Distributions &dist) {
  state.mid_price += dist.mid_walk(state.rng) + state.side_bias * 0.8;
  state.mid_price = std::max(state.mid_price, 100.0); // keep prices positive
}

// Drain every cancel scheduled for exactly this tick. Order-independent
// relative to the other phases -- it only touches ids from past ticks.
void fire_due_cancels(SimState &state, std::ofstream &out, std::uint64_t tick) {
  while (!state.pending_cancels.empty() &&
         state.pending_cancels.top().fire_at_tick == tick) {
    emit_cancel(out, state.pending_cancels.top().id, state.timestamp++);
    state.pending_cancels.pop();
  }
}

// Generate this tick's new order arrivals: how many, then for each one,
// side/type/size/price, emit it, and maybe schedule a future cancel.
void generate_arrivals(SimState &state, Distributions &dist, std::ofstream &out,
                       std::uint64_t tick) {
  int n_arrivals = (state.regime == Regime::Quiet)
                       ? dist.quiet_arrivals(state.rng)
                       : dist.burst_arrivals(state.rng);

  double p_buy = 0.5 + state.side_bias * 0.5; // bias skews buy/sell split
  p_buy = std::clamp(p_buy, 0.05, 0.95);
  std::bernoulli_distribution buy_dist(p_buy);

  for (int i = 0; i < n_arrivals; ++i) {
    OrderId id = state.next_id++;
    bool buy = buy_dist(state.rng);
    bool market = dist.is_market(state.rng);

    Qty qty =
        static_cast<Qty>(std::max(1.0, std::round(dist.size_dist(state.rng))));

    Price price = 0;
    if (!market) {
      Price offset = sample_laplace_offset(state.rng, /*scale=*/4.0);
      // Bias resting limit orders to sit on their own side of the mid
      // (buys below mid, sells above mid) so the book doesn't self-cross
      // on arrival more than realistically expected.
      Price signed_offset = buy ? -std::abs(offset) : std::abs(offset);
      price = static_cast<Price>(std::llround(state.mid_price)) + signed_offset;
      price = std::max<Price>(price, 1);
    }

    emit_new(out, id, buy, market, price, qty, state.timestamp++);

    if (!market && dist.will_cancel(state.rng)) {
      std::uint64_t fire_at =
          tick + static_cast<std::uint64_t>(dist.cancel_delay(state.rng));
      state.pending_cancels.push(PendingCancel{id, fire_at});
    }
  }
}

// Flush any cancels still pending past the last tick (their fire_at_tick
// exceeded num_ticks), so scheduled cancels aren't silently dropped from
// the dataset. The heap pops them in fire_at_tick order.
void flush_remaining_cancels(SimState &state, std::ofstream &out) {
  while (!state.pending_cancels.empty()) {
    emit_cancel(out, state.pending_cancels.top().id, state.timestamp++);
    state.pending_cancels.pop();
  }
}

int main(int argc, char **argv) {
  if (argc < 3) {
    std::cerr << "Usage: " << argv[0] << " <num_ticks> <output.csv> [seed]\n";
    return 1;
  }

  const std::uint64_t num_ticks = std::stoull(argv[1]);
  const std::string output_path = argv[2];
  const std::uint64_t seed = (argc >= 4) ? std::stoull(argv[3]) : 42ULL;

  std::ofstream out(output_path);
  if (!out) {
    std::cerr << "Failed to open output file: " << output_path << "\n";
    return 1;
  }
  out << "op,id,side,type,price,qty,timestamp\n";

  SimState state(seed);
  Distributions dist;

  for (std::uint64_t tick = 0; tick < num_ticks; ++tick) {
    update_regime(state, dist);
    update_mid_price(state, dist);
    fire_due_cancels(state, out, tick);
    generate_arrivals(state, dist, out, tick);
  }

  flush_remaining_cancels(state, out);

  std::cerr << "Generated " << (state.next_id - 1) << " orders across "
            << num_ticks << " ticks -> " << output_path << " (seed=" << seed
            << ")\n";
  return 0;
}